In [1]:
import pandas as pd
import numpy as np
 

In [2]:
#Create a DataFrame
d = {
    'Name':['Alisa','Bobby','jodha','jack','raghu','Cathrine',
            'Alisa','Bobby','kumar','Alisa','Alex','Cathrine'],
    'Age':[26,24,23,22,23,24,26,24,22,23,24,24],
      
    'Score':[85,63,55,74,31,77,85,63,42,62,89,77]}
 
df = pd.DataFrame(d)
df

Name  Age  Score
0      Alisa   26     85
1      Bobby   24     63
2      jodha   23     55
3       jack   22     74
4      raghu   23     31
5   Cathrine   24     77
6      Alisa   26     85
7      Bobby   24     63
8      kumar   22     42
9      Alisa   23     62
10      Alex   24     89
11  Cathrine   24     77

# Duplicates

#### Find the duplicate row in pandas:
duplicated() function is used for find the duplicate rows of the dataframe in python pandas

In [3]:

df.duplicated()

0     False
1     False
2     False
3     False
4     False
5     False
6      True
7      True
8     False
9     False
10    False
11     True
dtype: bool

The above code finds whether the row is duplicate and tags TRUE if it is duplicate and tags FALSE if it is not duplicate. 

And assigns it to the column named “is_duplicate”  of the dataframe  df.

In [4]:
df.duplicated("Name")


0     False
1     False
2     False
3     False
4     False
5     False
6      True
7      True
8     False
9      True
10    False
11     True
dtype: bool

This checks if there are duplicate values in a particular column of your DataFrame.

#### Getting rid of duplicates

In [5]:
df.drop_duplicates()

Name  Age  Score
0      Alisa   26     85
1      Bobby   24     63
2      jodha   23     55
3       jack   22     74
4      raghu   23     31
5   Cathrine   24     77
8      kumar   22     42
9      Alisa   23     62
10      Alex   24     89

In [6]:
# Sometimes, you may want to drop duplicates just from one column.


df.drop_duplicates(['Name'])
#df.drop_duplicates['Name']

Name  Age  Score
0      Alisa   26     85
1      Bobby   24     63
2      jodha   23     55
3       jack   22     74
4      raghu   23     31
5   Cathrine   24     77
8      kumar   22     42
10      Alex   24     89

# checking basic sanity of data

In [7]:
from sklearn import datasets

In [8]:
iris = datasets.load_iris()


In [9]:
cancer = datasets.load_breast_cancer()

In [10]:
iris.target

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

In [11]:
cancer.data

array([[1.799e+01, 1.038e+01, 1.228e+02, ..., 2.654e-01, 4.601e-01,
        1.189e-01],
       [2.057e+01, 1.777e+01, 1.329e+02, ..., 1.860e-01, 2.750e-01,
        8.902e-02],
       [1.969e+01, 2.125e+01, 1.300e+02, ..., 2.430e-01, 3.613e-01,
        8.758e-02],
       ...,
       [1.660e+01, 2.808e+01, 1.083e+02, ..., 1.418e-01, 2.218e-01,
        7.820e-02],
       [2.060e+01, 2.933e+01, 1.401e+02, ..., 2.650e-01, 4.087e-01,
        1.240e-01],
       [7.760e+00, 2.454e+01, 4.792e+01, ..., 0.000e+00, 2.871e-01,
        7.039e-02]])

In [12]:
cancer.target

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0,
       1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0,
       1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1,
       1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0,
       0, 1, 0, 0, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 0, 1, 1, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 0, 1, 1, 1,
       1, 0, 1, 1, 0, 0, 0, 1, 0, 1, 0, 1, 1, 1, 0, 1, 1, 0, 0, 1, 0, 0,
       0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 1, 1, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0,
       1, 1, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 1, 0, 1, 1, 0, 0, 1, 0, 1, 1,
       1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 0, 1, 1, 0, 1, 0, 0, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0,

In [13]:
print(iris.DESCR)

.. _iris_dataset:

Iris plants dataset
--------------------

**Data Set Characteristics:**

    :Number of Instances: 150 (50 in each of three classes)
    :Number of Attributes: 4 numeric, predictive attributes and the class
    :Attribute Information:
        - sepal length in cm
        - sepal width in cm
        - petal length in cm
        - petal width in cm
        - class:
                - Iris-Setosa
                - Iris-Versicolour
                - Iris-Virginica
                
    :Summary Statistics:

    ============== ==== ==== ======= ===== ====================
                    Min  Max   Mean    SD   Class Correlation
    ============== ==== ==== ======= ===== ====================
    sepal length:   4.3  7.9   5.84   0.83    0.7826
    sepal width:    2.0  4.4   3.05   0.43   -0.4194
    petal length:   1.0  6.9   3.76   1.76    0.9490  (high!)
    petal width:    0.1  2.5   1.20   0.76    0.9565  (high!)
    ============== ==== ==== ======= ===== ===========

In [16]:
# iris['sepal length'].agg(['mean','max','min','count','median','std','var'])

In [17]:
cancer.target_names

array(['malignant', 'benign'], dtype='<U9')

In [18]:
iris.target_names

array(['setosa', 'versicolor', 'virginica'], dtype='<U10')

In [19]:
X = iris.data
y = iris.target

In [20]:
X[:2]

array([[5.1, 3.5, 1.4, 0.2],
       [4.9, 3. , 1.4, 0.2]])

In [21]:
X[:4]

array([[5.1, 3.5, 1.4, 0.2],
       [4.9, 3. , 1.4, 0.2],
       [4.7, 3.2, 1.3, 0.2],
       [4.6, 3.1, 1.5, 0.2]])

In [22]:
y[:5]

array([0, 0, 0, 0, 0])

In [23]:
# np.c_[X[:5], y[:5]]

np.c_[X[:6], y[:6]]

array([[5.1, 3.5, 1.4, 0.2, 0. ],
       [4.9, 3. , 1.4, 0.2, 0. ],
       [4.7, 3.2, 1.3, 0.2, 0. ],
       [4.6, 3.1, 1.5, 0.2, 0. ],
       [5. , 3.6, 1.4, 0.2, 0. ],
       [5.4, 3.9, 1.7, 0.4, 0. ]])

In [24]:
# data    = np.c_[X, y]
data = np.c_[X, y]
# data = np.c_[X, y]
# data.np.c_[X,y]
# data =np.c_[X,y]
# data = np.c_[X,y]

In [25]:

column = ['sepal_Length', 'sepal_Width', 'petal_Length', 'petal_Width', 'class']

# df      = pd.DataFrame(data, columns=column)
# df.head()
# df=pd.DataFrame(data, columns=column)
# df.head()
df=pd.DataFrame(data, columns=column)
df.head(4)

sepal_Length  sepal_Width  petal_Length  petal_Width  class
0           5.1          3.5           1.4          0.2    0.0
1           4.9          3.0           1.4          0.2    0.0
2           4.7          3.2           1.3          0.2    0.0
3           4.6          3.1           1.5          0.2    0.0

In [26]:
df['sepal_Length'].agg(['mean','max','min','count','median','std','var'])

mean        5.843333
max         7.900000
min         4.300000
count     150.000000
median      5.800000
std         0.828066
var         0.685694
Name: sepal_Length, dtype: float64

In [23]:
# change the class from numeric to string for easy understanding
replacements  = {0: 'setosa', 1: 'versicolor', 2: 'virginica'}

# df['class']   = df['class'].map(replacements)

In [24]:
# df['class']=df['class'].map(replacements)
df['class_rep'] = df['class'].map(replacements)

In [25]:
df.head()

sepal_Length  sepal_Width  petal_Length  petal_Width  class class_rep
0           5.1          3.5           1.4          0.2    0.0    setosa
1           4.9          3.0           1.4          0.2    0.0    setosa
2           4.7          3.2           1.3          0.2    0.0    setosa
3           4.6          3.1           1.5          0.2    0.0    setosa
4           5.0          3.6           1.4          0.2    0.0    setosa

In [26]:
# # # find out how many samples from each class
# # df['class'].value_counts()
# df['class'].value_counts()
df['class'].value_counts()

2.0    50
1.0    50
0.0    50
Name: class, dtype: int64

In [27]:
# # basic statistics
# # check for min, max, quartiles, IQR
# df.describe()
# df.describe()
df.describe()

sepal_Length  sepal_Width  petal_Length  petal_Width       class
count    150.000000   150.000000    150.000000   150.000000  150.000000
mean       5.843333     3.057333      3.758000     1.199333    1.000000
std        0.828066     0.435866      1.765298     0.762238    0.819232
min        4.300000     2.000000      1.000000     0.100000    0.000000
25%        5.100000     2.800000      1.600000     0.300000    0.000000
50%        5.800000     3.000000      4.350000     1.300000    1.000000
75%        6.400000     3.300000      5.100000     1.800000    2.000000
max        7.900000     4.400000      6.900000     2.500000    2.000000

In [28]:
df[['sepal_Length', 'petal_Length']].describe()

sepal_Length  petal_Length
count    150.000000    150.000000
mean       5.843333      3.758000
std        0.828066      1.765298
min        4.300000      1.000000
25%        5.100000      1.600000
50%        5.800000      4.350000
75%        6.400000      5.100000
max        7.900000      6.900000

In [29]:
# #calculate the mean petal length of each of the three classes
#df.loc[df['class'] == 'setosa', metrics].mean(axis=0)

In [30]:
# # returns indexes
#  df['class'] == 'setosa'

In [31]:
# locate by index position
# df.loc[[0,6,5]]
# df.loc[[1,10,15]]
df.loc[[1,10,5,8]]

sepal_Length  sepal_Width  petal_Length  petal_Width  class class_rep
1            4.9          3.0           1.4          0.2    0.0    setosa
10           5.4          3.7           1.5          0.2    0.0    setosa
5            5.4          3.9           1.7          0.4    0.0    setosa
8            4.4          2.9           1.4          0.2    0.0    setosa

In [32]:
# # df.loc[df['class'] == 'setosa']
# df.loc[df['class'] == 'versicolor']

df.loc[df['class_rep'] == 'setosa'].head(5)

sepal_Length  sepal_Width  petal_Length  petal_Width  class class_rep
0           5.1          3.5           1.4          0.2    0.0    setosa
1           4.9          3.0           1.4          0.2    0.0    setosa
2           4.7          3.2           1.3          0.2    0.0    setosa
3           4.6          3.1           1.5          0.2    0.0    setosa
4           5.0          3.6           1.4          0.2    0.0    setosa

In [33]:
# display selected columns for class = 'setosa'
# df.loc[df['class'] == 'setosa', ['petal_Width', 'petal_Length']].head()

# df.loc[df['class_rep'] == 'versicolor' , ['petal_Width' , 'petal_Length']]
df.loc[df['class_rep'] == 'setosa' , ['petal_Width', 'petal_Length']].head(5)

petal_Width  petal_Length
0          0.2           1.4
1          0.2           1.4
2          0.2           1.3
3          0.2           1.5
4          0.2           1.4

In [34]:
# display mean values of selected columns for class = 'setosa'
df.loc[df['class_rep'] == 'setosa', ['petal_Width', 'petal_Length']].mean()

petal_Width     0.246
petal_Length    1.462
dtype: float64

In [35]:
# display mean values of selected columns for class = 'setosa'
df.loc[df['class_rep'] == 'setosa', ['petal_Width', 'petal_Length']].max()

petal_Width     0.6
petal_Length    1.9
dtype: float64

In [37]:
df.describe()

sepal_Length  sepal_Width  petal_Length  petal_Width       class
count    150.000000   150.000000    150.000000   150.000000  150.000000
mean       5.843333     3.057333      3.758000     1.199333    1.000000
std        0.828066     0.435866      1.765298     0.762238    0.819232
min        4.300000     2.000000      1.000000     0.100000    0.000000
25%        5.100000     2.800000      1.600000     0.300000    0.000000
50%        5.800000     3.000000      4.350000     1.300000    1.000000
75%        6.400000     3.300000      5.100000     1.800000    2.000000
max        7.900000     4.400000      6.900000     2.500000    2.000000

In [38]:
df.loc[ df['class_rep']=='versicolor' , ['petal_Length', 'petal_Width']].mean()

petal_Length    4.260
petal_Width     1.326
dtype: float64

In [39]:
df.loc[ df['class_rep']=='versicolor' , ['petal_Length', 'petal_Width']].count()

petal_Length    50
petal_Width     50
dtype: int64

In [40]:
df['class_rep'].value_counts()

virginica     50
setosa        50
versicolor    50
Name: class_rep, dtype: int64

In [41]:
# use of grouping
groups = df.groupby('class')

In [42]:
grp = df.groupby('petal_Width')

In [43]:
#grp.mean()

In [44]:
# use sum(), min(), max(), count()
groups.mean()

sepal_Length  sepal_Width  petal_Length  petal_Width
class                                                      
0.0           5.006        3.428         1.462        0.246
1.0           5.936        2.770         4.260        1.326
2.0           6.588        2.974         5.552        2.026

In [45]:
groups.agg(['mean', 'min', 'max', 'count'])

sepal_Length                 sepal_Width                 petal_Length  \
              mean  min  max count        mean  min  max count         mean   
class                                                                         
0.0          5.006  4.3  5.8    50       3.428  2.3  4.4    50        1.462   
1.0          5.936  4.9  7.0    50       2.770  2.0  3.4    50        4.260   
2.0          6.588  4.9  7.9    50       2.974  2.2  3.8    50        5.552   

                      petal_Width                  
       min  max count        mean  min  max count  
class                                              
0.0    1.0  1.9    50       0.246  0.1  0.6    50  
1.0    3.0  5.1    50       1.326  1.0  1.8    50  
2.0    4.5  6.9    50       2.026  1.4  2.5    50

In [46]:
# grp.agg(['mean','min','max','count'])

In [47]:
df.head()

sepal_Length  sepal_Width  petal_Length  petal_Width  class class_rep
0           5.1          3.5           1.4          0.2    0.0    setosa
1           4.9          3.0           1.4          0.2    0.0    setosa
2           4.7          3.2           1.3          0.2    0.0    setosa
3           4.6          3.1           1.5          0.2    0.0    setosa
4           5.0          3.6           1.4          0.2    0.0    setosa

# numeric col containing non-numeric data

In [48]:
d = {
 'unit': ['UD', 'UD', 'UD', 'UD', 'ud','uD'],
 'N-D':  [ 'Q1', 'Q2', 'Q3', 'Q4','Q5','Q6'],
 'num' : [ -1.48, 1.7, -6.18, 0.25, 'weird', 0.25]

}

dff = pd.DataFrame(d)
dff

unit N-D    num
0   UD  Q1  -1.48
1   UD  Q2    1.7
2   UD  Q3  -6.18
3   UD  Q4   0.25
4   ud  Q5  weird
5   uD  Q6   0.25

In [49]:
type(df)

pandas.core.frame.DataFrame

In [50]:
type(dff)

pandas.core.frame.DataFrame

In [51]:
dff['num'].str.istitle()

0      NaN
1      NaN
2      NaN
3      NaN
4    False
5      NaN
Name: num, dtype: object

In [52]:
dff['num'].str.isdecimal()

0      NaN
1      NaN
2      NaN
3      NaN
4    False
5      NaN
Name: num, dtype: object

# check for character cases

In [53]:
dff['unit'].str.isupper()


0     True
1     True
2     True
3     True
4    False
5    False
Name: unit, dtype: bool

In [54]:
dff['unit'] = dff['unit'].str.upper()

In [55]:
dff

unit N-D    num
0   UD  Q1  -1.48
1   UD  Q2    1.7
2   UD  Q3  -6.18
3   UD  Q4   0.25
4   UD  Q5  weird
5   UD  Q6   0.25

# handling missing values

In [56]:
location = r"F:\EducationalData\Simplilearn\ML\Files\datasets\nba.csv"

In [57]:
# making data frame from csv file 
nba = pd.read_csv(location)

In [58]:
nba.shape

(457, 9)

In [59]:
nba.head(5)

Name            Team  Number Position  Age  Height  Weight  \
0  Avery Bradley  Boston Celtics       0       PG   25   2-Jun     180   
1    Jae Crowder  Boston Celtics      99       SF   25   6-Jun     235   
2   John Holland  Boston Celtics      30       SG   27   5-Jun     205   
3    R.J. Hunter  Boston Celtics      28       SG   22   5-Jun     185   
4  Jonas Jerebko  Boston Celtics       8       PF   29  10-Jun     231   

             College     Salary  
0              Texas  7730337.0  
1          Marquette  6796117.0  
2  Boston University        NaN  
3      Georgia State  1148640.0  
4                NaN  5000000.0

In [60]:
# are there nulls?
# nba.isnull().sum()
# nba.isnull()
nba.isnull().sum()

Name         0
Team         0
Number       0
Position     0
Age          0
Height       0
Weight       0
College     84
Salary      11
dtype: int64

In [61]:
nba.count()

Name        457
Team        457
Number      457
Position    457
Age         457
Height      457
Weight      457
College     373
Salary      446
dtype: int64

example #1 - Replacing NaN values with a Static value.

In [62]:
# # replacing na values in college with No college 
# nba["College"].fillna("No College", inplace = True) 
# nba.isnull().sum()
# #nba['College'].isnull().sum()

nba['College'].fillna('No College', inplace = True)
nba.isnull().sum()

Name         0
Team         0
Number       0
Position     0
Age          0
Height       0
Weight       0
College      0
Salary      11
dtype: int64

In [63]:
nba['Salary'].fillna(0, inplace=True)
nba.isnull().sum()

Name        0
Team        0
Number      0
Position    0
Age         0
Height      0
Weight      0
College     0
Salary      0
dtype: int64

Example #2: Using method Parameter

In [64]:
# making data frame from csv file 
nba = pd.read_csv(location) 
nba.head()

Name            Team  Number Position  Age  Height  Weight  \
0  Avery Bradley  Boston Celtics       0       PG   25   2-Jun     180   
1    Jae Crowder  Boston Celtics      99       SF   25   6-Jun     235   
2   John Holland  Boston Celtics      30       SG   27   5-Jun     205   
3    R.J. Hunter  Boston Celtics      28       SG   22   5-Jun     185   
4  Jonas Jerebko  Boston Celtics       8       PF   29  10-Jun     231   

             College     Salary  
0              Texas  7730337.0  
1          Marquette  6796117.0  
2  Boston University        NaN  
3      Georgia State  1148640.0  
4                NaN  5000000.0

In [65]:
nba = pd.read_csv(location) 
nba['College'].describe()

count          373
unique         118
top       Kentucky
freq            22
Name: College, dtype: object

In [66]:
# making data frame from csv file 
nba = pd.read_csv(location) 
  
# replacing na values in college with previous value college 
nba["College"].fillna( method ='ffill', inplace = True) 
  
nba.head()

Name            Team  Number Position  Age  Height  Weight  \
0  Avery Bradley  Boston Celtics       0       PG   25   2-Jun     180   
1    Jae Crowder  Boston Celtics      99       SF   25   6-Jun     235   
2   John Holland  Boston Celtics      30       SG   27   5-Jun     205   
3    R.J. Hunter  Boston Celtics      28       SG   22   5-Jun     185   
4  Jonas Jerebko  Boston Celtics       8       PF   29  10-Jun     231   

             College     Salary  
0              Texas  7730337.0  
1          Marquette  6796117.0  
2  Boston University        NaN  
3      Georgia State  1148640.0  
4      Georgia State  5000000.0

Similarly, bfill, backfill and pad methods can also be used.

Example #3: Using mean value

In [67]:
# making data frame from csv file 
nba = pd.read_csv(location) 
nba.head(10)

Name            Team  Number Position  Age  Height  Weight  \
0  Avery Bradley  Boston Celtics       0       PG   25   2-Jun     180   
1    Jae Crowder  Boston Celtics      99       SF   25   6-Jun     235   
2   John Holland  Boston Celtics      30       SG   27   5-Jun     205   
3    R.J. Hunter  Boston Celtics      28       SG   22   5-Jun     185   
4  Jonas Jerebko  Boston Celtics       8       PF   29  10-Jun     231   
5   Amir Johnson  Boston Celtics      90       PF   29   9-Jun     240   
6  Jordan Mickey  Boston Celtics      55       PF   21   8-Jun     235   
7   Kelly Olynyk  Boston Celtics      41        C   25  Jul-00     238   
8   Terry Rozier  Boston Celtics      12       PG   22   2-Jun     190   
9   Marcus Smart  Boston Celtics      36       PG   22   4-Jun     220   

             College      Salary  
0              Texas   7730337.0  
1          Marquette   6796117.0  
2  Boston University         NaN  
3      Georgia State   1148640.0  
4                NaN   5000000.0  
5                NaN  12000000.0  
6                LSU   1170960.0  
7            Gonzaga   2165160.0  
8         Louisville   1824360.0  
9     Oklahoma State   3431040.0

In [68]:
nba.isnull().sum()

Name         0
Team         0
Number       0
Position     0
Age          0
Height       0
Weight       0
College     84
Salary      11
dtype: int64

In [69]:
# # Or, depending on your needs:
# #nba['Salary'] = nba['Salary'].fillna(nba['Salary'].mean())
# nba['Salary'] = nba['Salary'].fillna(nba['Salary'].min())
# #nba['Salary'] = nba['Salary'].fillna(nba['Salary'].max())
# nba.head(10)
# nba['Salary'] = nba['Salary'].fillna(nba['Salary'].min)
nba['Salary'] = nba['Salary'].fillna(nba['Salary'].min)
nba.head()

Name            Team  Number Position  Age  Height  Weight  \
0  Avery Bradley  Boston Celtics       0       PG   25   2-Jun     180   
1    Jae Crowder  Boston Celtics      99       SF   25   6-Jun     235   
2   John Holland  Boston Celtics      30       SG   27   5-Jun     205   
3    R.J. Hunter  Boston Celtics      28       SG   22   5-Jun     185   
4  Jonas Jerebko  Boston Celtics       8       PF   29  10-Jun     231   

             College                                             Salary  
0              Texas                                        7.73034e+06  
1          Marquette                                        6.79612e+06  
2  Boston University  <bound method Series.min of 0       7730337.0\...  
3      Georgia State                                        1.14864e+06  
4                NaN                                              5e+06

In [70]:
nba.isnull().sum()

Name         0
Team         0
Number       0
Position     0
Age          0
Height       0
Weight       0
College     84
Salary       0
dtype: int64

In [71]:
nba['Age'] = nba['Age'].fillna(200)

In [72]:
nba['Age'].isnull().sum()

0

In [73]:
# nba[nba['Age']==25]

In [74]:
raw_data = {'first_name': ['Jason', np.nan, 'Tina', 'Jake', 'Amy'], 
        'last_name': ['Miller', np.nan, np.nan, 'Milner', 'Cooze'], 
        'age': [42, np.nan, 36, 24, 73], 
        'sex': ['m', np.nan, 'f', 'm', 'f'], 
        'preTestScore': [4, np.nan, np.nan, 2, 3],
        'postTestScore': [25, np.nan, np.nan, 62, 70]}
df = pd.DataFrame(raw_data, columns = ['first_name', 'last_name', 'age', 'sex', 'preTestScore', 'postTestScore'])
df

first_name last_name   age  sex  preTestScore  postTestScore
0      Jason    Miller  42.0    m           4.0           25.0
1        NaN       NaN   NaN  NaN           NaN            NaN
2       Tina       NaN  36.0    f           NaN            NaN
3       Jake    Milner  24.0    m           2.0           62.0
4        Amy     Cooze  73.0    f           3.0           70.0

In [75]:
# Drop missing observations
# rows returned after dropping, the original df remains intact
df_no_missing = df.dropna()
df_no_missing.head(5)

first_name last_name   age sex  preTestScore  postTestScore
0      Jason    Miller  42.0   m           4.0           25.0
3       Jake    Milner  24.0   m           2.0           62.0
4        Amy     Cooze  73.0   f           3.0           70.0

In [76]:
df_no_missing.head(5)

first_name last_name   age sex  preTestScore  postTestScore
0      Jason    Miller  42.0   m           4.0           25.0
3       Jake    Milner  24.0   m           2.0           62.0
4        Amy     Cooze  73.0   f           3.0           70.0

In [77]:
df_no_missing.isnull().sum()

first_name       0
last_name        0
age              0
sex              0
preTestScore     0
postTestScore    0
dtype: int64

In [81]:
#Drop rows where all cells in that row is NA
df_cleaned = df.dropna(how='all')
df_cleaned

first_name last_name   age sex  preTestScore  postTestScore  location
0      Jason    Miller  42.0   m           4.0           25.0       NaN
2       Tina       NaN  36.0   f           NaN            NaN       NaN
3       Jake    Milner  24.0   m           2.0           62.0       NaN
4        Amy     Cooze  73.0   f           3.0           70.0       NaN

In [82]:
# Create a new column full of missing values
df['location'] = np.nan
df

first_name last_name   age  sex  preTestScore  postTestScore  location
0      Jason    Miller  42.0    m           4.0           25.0       NaN
1        NaN       NaN   NaN  NaN           NaN            NaN       NaN
2       Tina       NaN  36.0    f           NaN            NaN       NaN
3       Jake    Milner  24.0    m           2.0           62.0       NaN
4        Amy     Cooze  73.0    f           3.0           70.0       NaN

In [85]:
# Drop column if they only contain missing values
df.dropna(axis=0, how='all')

first_name last_name   age sex  preTestScore  postTestScore  location
0      Jason    Miller  42.0   m           4.0           25.0       NaN
2       Tina       NaN  36.0   f           NaN            NaN       NaN
3       Jake    Milner  24.0   m           2.0           62.0       NaN
4        Amy     Cooze  73.0   f           3.0           70.0       NaN

In [86]:
df.head()

first_name last_name   age  sex  preTestScore  postTestScore  location
0      Jason    Miller  42.0    m           4.0           25.0       NaN
1        NaN       NaN   NaN  NaN           NaN            NaN       NaN
2       Tina       NaN  36.0    f           NaN            NaN       NaN
3       Jake    Milner  24.0    m           2.0           62.0       NaN
4        Amy     Cooze  73.0    f           3.0           70.0       NaN

In [87]:
# Drop rows that contain less than five observations
# This is really mostly useful for time series
df
df.dropna(thresh=4, axis=0).head()

first_name last_name   age sex  preTestScore  postTestScore  location
0      Jason    Miller  42.0   m           4.0           25.0       NaN
3       Jake    Milner  24.0   m           2.0           62.0       NaN
4        Amy     Cooze  73.0   f           3.0           70.0       NaN

In [88]:
# Fill in missing data with zeros
df.fillna(0)

first_name last_name   age sex  preTestScore  postTestScore  location
0      Jason    Miller  42.0   m           4.0           25.0       0.0
1          0         0   0.0   0           0.0            0.0       0.0
2       Tina         0  36.0   f           0.0            0.0       0.0
3       Jake    Milner  24.0   m           2.0           62.0       0.0
4        Amy     Cooze  73.0   f           3.0           70.0       0.0

In [89]:
# Fill in missing in preTestScore with the mean value of preTestScore
# inplace=True means that the changes are saved to the df right away

In [91]:
df["preTestScore"].fillna(df["preTestScore"].mean(), inplace=True)
df

first_name last_name   age  sex  preTestScore  postTestScore  location
0      Jason    Miller  42.0    m           4.0           25.0       NaN
1        NaN       NaN   NaN  NaN           3.0            NaN       NaN
2       Tina       NaN  36.0    f           3.0            NaN       NaN
3       Jake    Milner  24.0    m           2.0           62.0       NaN
4        Amy     Cooze  73.0    f           3.0           70.0       NaN

In [94]:
# Fill in missing in postTestScore with each sex’s mean value of postTestScore
df["postTestScore"].fillna(df.groupby("sex")["postTestScore"].transform("mean"), inplace=True)
df
# df['postTestScore'].fillna(df.groupby('sex')['postTestScore'].transform('mean'), inplace=True)
# df

first_name last_name   age  sex  preTestScore  postTestScore  location
0      Jason    Miller  42.0    m           4.0           25.0       NaN
1        NaN       NaN   NaN  NaN           3.0            NaN       NaN
2       Tina       NaN  36.0    f           3.0           70.0       NaN
3       Jake    Milner  24.0    m           2.0           62.0       NaN
4        Amy     Cooze  73.0    f           3.0           70.0       NaN

#### using scikit learn imputer

strategy : string, optional (default=”mean”)
The imputation strategy.

- If “mean”, then replace missing values using the mean along each column. Can only be used with numeric data.
- If “median”, then replace missing values using the median along each column. Can only be used with numeric data.
- If “most_frequent”, then replace missing using the most frequent value along each column. Can be used with strings or numeric data.
- If “constant”, then replace missing values with fill_value. Can be used with strings or numeric data.

In [95]:
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import Imputer

df = pd.DataFrame([[10,     20],
                   [np.nan, 23],
                   [15,    np.nan],
                   [17,    29]])

df.dtypes

0    float64
1    float64
dtype: object

In [96]:
imp = SimpleImputer(strategy="most_frequent")
print(imp.fit_transform(df)) 

[[10. 20.]
 [10. 23.]
 [15. 20.]
 [17. 29.]]


one more example ...

In [97]:
X = np.array([[23.56],
              [53.45],
              ['NaN'],
              [44.44],
              [77.78],
              ['NaN'],
              [234.44],
              [11.33],
              [79.87]])

In [98]:
imputer = SimpleImputer(strategy='mean')  

imputer.fit_transform(X)

array([[ 23.56      ],
       [ 53.45      ],
       [ 74.98142857],
       [ 44.44      ],
       [ 77.78      ],
       [ 74.98142857],
       [234.44      ],
       [ 11.33      ],
       [ 79.87      ]])

one more example ..

In [99]:
# Create an empty dataset
df = pd.DataFrame()

# Create two variables called x0 and x1. Make the first value of x1 a missing value
df['x0'] = [0.3051,0.4949,0.6974,0.3769,0.2231,0.341,0.4436,0.5897,0.6308,0.5]
df['x1'] = [np.nan,0.2654,0.2615,0.5846,0.4615,0.8308,0.4962,0.3269,0.5346,0.6731]

# View the dataset
df

x0      x1
0  0.3051     NaN
1  0.4949  0.2654
2  0.6974  0.2615
3  0.3769  0.5846
4  0.2231  0.4615
5  0.3410  0.8308
6  0.4436  0.4962
7  0.5897  0.3269
8  0.6308  0.5346
9  0.5000  0.6731

In [100]:
# Fit Imputer
# Create an imputer object that looks for 'Nan' values, then replaces them with the mean value of the feature by columns (axis=0)
mean_imputer = Imputer(missing_values='NaN', strategy='mean', axis=0)

# Train the imputor on the df dataset
mean_imputer = mean_imputer.fit(df)

C:\Users\Satnam\Anaconda3\lib\site-packages\sklearn\utils\deprecation.py:58: DeprecationWarning: Class Imputer is deprecated; Imputer was deprecated in version 0.20 and will be removed in 0.22. Import impute.SimpleImputer from sklearn instead.
  warnings.warn(msg, category=DeprecationWarning)


In [101]:
# Apply Imputer
# Apply the imputer to the df dataset
imputed_df = mean_imputer.transform(df.values)

In [102]:
imputed_df

array([[0.3051    , 0.49273333],
       [0.4949    , 0.2654    ],
       [0.6974    , 0.2615    ],
       [0.3769    , 0.5846    ],
       [0.2231    , 0.4615    ],
       [0.341     , 0.8308    ],
       [0.4436    , 0.4962    ],
       [0.5897    , 0.3269    ],
       [0.6308    , 0.5346    ],
       [0.5       , 0.6731    ]])

Notice that 0.49273333 is the imputed value, replacing the np.NaN value.

#### more examples ... multiple columns

In [103]:
titanic = pd.read_csv("https://raw.githubusercontent.com/amueller/scipy-2017-sklearn/master/notebooks/datasets/titanic3.csv")

In [104]:
titanic.describe()

pclass     survived          age        sibsp        parch  \
count  1309.000000  1309.000000  1046.000000  1309.000000  1309.000000   
mean      2.294882     0.381971    29.881135     0.498854     0.385027   
std       0.837836     0.486055    14.413500     1.041658     0.865560   
min       1.000000     0.000000     0.166700     0.000000     0.000000   
25%       2.000000     0.000000    21.000000     0.000000     0.000000   
50%       3.000000     0.000000    28.000000     0.000000     0.000000   
75%       3.000000     1.000000    39.000000     1.000000     0.000000   
max       3.000000     1.000000    80.000000     8.000000     9.000000   

              fare        body  
count  1308.000000  121.000000  
mean     33.295479  160.809917  
std      51.758668   97.696922  
min       0.000000    1.000000  
25%       7.895800   72.000000  
50%      14.454200  155.000000  
75%      31.275000  256.000000  
max     512.329200  328.000000

In [105]:
#titanic = pd.read_csv("https://raw.githubusercontent.com/amueller/scipy-2017-sklearn/master/notebooks/datasets/titanic3.csv")
# there is still a small problem with using the OneHotEncoder and missing values,
# so for now I am going to assume there are no missing values by dropping them

titanic2 = titanic.dropna(subset=['pclass', 'sex', 'age', 'sibsp', 'parch', 'fare', 'embarked'])

In [106]:
titanic.head()

pclass  survived                                             name     sex  \
0       1         1                    Allen, Miss. Elisabeth Walton  female   
1       1         1                   Allison, Master. Hudson Trevor    male   
2       1         0                     Allison, Miss. Helen Loraine  female   
3       1         0             Allison, Mr. Hudson Joshua Creighton    male   
4       1         0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  female   

       age  sibsp  parch  ticket      fare    cabin embarked boat   body  \
0  29.0000      0      0   24160  211.3375       B5        S    2    NaN   
1   0.9167      1      2  113781  151.5500  C22 C26        S   11    NaN   
2   2.0000      1      2  113781  151.5500  C22 C26        S  NaN    NaN   
3  30.0000      1      2  113781  151.5500  C22 C26        S  NaN  135.0   
4  25.0000      1      2  113781  151.5500  C22 C26        S  NaN    NaN   

                         home.dest  
0                     St Louis, MO  
1  Montreal, PQ / Chesterville, ON  
2  Montreal, PQ / Chesterville, ON  
3  Montreal, PQ / Chesterville, ON  
4  Montreal, PQ / Chesterville, ON

In [107]:
titanic2.head()

pclass  survived                                             name     sex  \
0       1         1                    Allen, Miss. Elisabeth Walton  female   
1       1         1                   Allison, Master. Hudson Trevor    male   
2       1         0                     Allison, Miss. Helen Loraine  female   
3       1         0             Allison, Mr. Hudson Joshua Creighton    male   
4       1         0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  female   

       age  sibsp  parch  ticket      fare    cabin embarked boat   body  \
0  29.0000      0      0   24160  211.3375       B5        S    2    NaN   
1   0.9167      1      2  113781  151.5500  C22 C26        S   11    NaN   
2   2.0000      1      2  113781  151.5500  C22 C26        S  NaN    NaN   
3  30.0000      1      2  113781  151.5500  C22 C26        S  NaN  135.0   
4  25.0000      1      2  113781  151.5500  C22 C26        S  NaN    NaN   

                         home.dest  
0                     St Louis, MO  
1  Montreal, PQ / Chesterville, ON  
2  Montreal, PQ / Chesterville, ON  
3  Montreal, PQ / Chesterville, ON  
4  Montreal, PQ / Chesterville, ON

In [108]:
target   = titanic2.survived.values
features = titanic2[['pclass', 'sex', 'age', 'fare', 'embarked']]

In [109]:
features.head()

pclass     sex      age      fare embarked
0       1  female  29.0000  211.3375        S
1       1    male   0.9167  151.5500        S
2       1  female   2.0000  151.5500        S
3       1    male  30.0000  151.5500        S
4       1  female  25.0000  151.5500        S

In [110]:
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.compose import ColumnTransformer, make_column_transformer

In [111]:
preprocess = make_column_transformer(
    (StandardScaler(), ['age', 'fare']),
    (OneHotEncoder(),  ['pclass', 'sex', 'embarked'])
)

In [112]:
new_features = preprocess.fit_transform(features)
new_features[:3]

array([[-0.05663194,  3.13554913,  1.        ,  0.        ,  0.        ,
         1.        ,  0.        ,  0.        ,  0.        ,  1.        ],
       [-2.01237899,  2.06268333,  1.        ,  0.        ,  0.        ,
         0.        ,  1.        ,  0.        ,  0.        ,  1.        ],
       [-1.93693697,  2.06268333,  1.        ,  0.        ,  0.        ,
         1.        ,  0.        ,  0.        ,  0.        ,  1.        ]])

In [113]:
new_features_df = pd.DataFrame(data=new_features)
new_features_df.head()

0         1    2    3    4    5    6    7    8    9
0 -0.056632  3.135549  1.0  0.0  0.0  1.0  0.0  0.0  0.0  1.0
1 -2.012379  2.062683  1.0  0.0  0.0  0.0  1.0  0.0  0.0  1.0
2 -1.936937  2.062683  1.0  0.0  0.0  1.0  0.0  0.0  0.0  1.0
3  0.013009  2.062683  1.0  0.0  0.0  0.0  1.0  0.0  0.0  1.0
4 -0.335196  2.062683  1.0  0.0  0.0  1.0  0.0  0.0  0.0  1.0